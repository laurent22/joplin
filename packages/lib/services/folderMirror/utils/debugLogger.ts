import Logger, { LogLevel, TargetType } from '@joplin/utils/Logger';

const debugLogger = new Logger();
debugLogger.addTarget(TargetType.Console);
debugLogger.setLevel(LogLevel.Debug);
debugLogger.enabled = false;

export default debugLogger;
