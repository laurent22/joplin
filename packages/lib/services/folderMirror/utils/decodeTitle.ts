
const decodeTitle = (title: string) => decodeURIComponent(title);

export default decodeTitle;
