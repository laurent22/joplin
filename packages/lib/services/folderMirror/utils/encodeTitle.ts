
// Encodes `title` into a file-system-safe format that preserves as many characters
// as is safe.
const encodeTitle = (title: string) => {
	// See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent#encoding_for_rfc3986
	return title.replace(
		/[%!\\/)]/g,
		c => `%${c.charCodeAt(0).toString(16)}`,
	);
};

export default encodeTitle;
