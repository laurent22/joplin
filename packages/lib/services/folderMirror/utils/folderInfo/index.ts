
export const folderInfoFileName = '.folder.yml';

export interface FolderInfo {
	id?: string;
	title: string;
	icon: string;
	folder_info_updated?: number;
}
