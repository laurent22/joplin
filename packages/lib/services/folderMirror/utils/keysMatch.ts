import { FolderEntity, NoteEntity, ResourceEntity } from '../../database/types';
import { FolderItem } from '../types';

const keysMatch = (localItem: FolderItem, remoteItem: FolderItem, keys: ((keyof FolderEntity)|(keyof NoteEntity)|(keyof ResourceEntity))[]) => {
	for (const key of keys) {
		if (key in localItem !== key in remoteItem) {
			return false;
		}
		if (key in localItem && localItem[key as keyof typeof localItem] !== remoteItem[key as keyof typeof remoteItem]) {
			return false;
		}
	}
	return true;
};

export default keysMatch;
