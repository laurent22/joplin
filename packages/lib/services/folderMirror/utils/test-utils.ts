import * as fs from 'fs-extra';
import eventManager, { EventName, ItemChangeEvent } from '../../../eventManager';
import { ModelType } from '../../../BaseModel';
import Note from '../../../models/Note';
import { join } from 'path';
import { NoteEntity } from '../../database/types';
import { setupDatabaseAndSynchronizer, switchClient } from '../../../testing/test-utils';
import reducer, { defaultState } from '../../../reducer';
import { createStore } from 'redux';
import BaseItem from '../../../models/BaseItem';
import FolderMirroringService from '../FolderMirroringService';
const md5 = require('md5');


type ShouldMatchItemCallback = (item: NoteEntity)=> boolean;
export const waitForNoteChange = (itemMatcher?: ShouldMatchItemCallback) => {
	return new Promise<void>(resolve => {
		const onResolve = () => {
			eventManager.off(EventName.ItemChange, eventHandler);
			resolve();
		};

		const eventHandler = async (event: ItemChangeEvent) => {
			if (event.itemType !== ModelType.Note) return;

			if (!itemMatcher) {
				onResolve();
			} else if (itemMatcher(await Note.load(event.itemId))) {
				onResolve();
			}
		};

		eventManager.on(EventName.ItemChange, eventHandler);
	});
};

export const waitForTestNoteToBeWritten = async (parentDir: string) => {
	// Push a new writeFile task to the end of the action queue and wait for it.
	const waitForActionsToComplete = waitForNoteChange(item => item.body === 'waitForActionsToComplete');
	await fs.writeFile(join(parentDir, 'waitForQueue.md'), 'waitForActionsToComplete', 'utf8');
	await waitForActionsToComplete;

	const waitForDeleteAction = waitForNoteChange(item => item.body === 'waitForActionsToComplete');
	await fs.remove(join(parentDir, 'waitForQueue.md'));
	await waitForDeleteAction;
};

export const setUpFolderMirror = async () => {
	await setupDatabaseAndSynchronizer(1);
	await switchClient(1);

	const testReducer = (state = defaultState, action: unknown) => {
		return reducer(state, action);
	};
	const store = createStore(testReducer);

	BaseItem.dispatch = store.dispatch;

	// Needed for AsyncActionQueue
	jest.useRealTimers();
};

export const resetFolderMirror = async () => {
	await FolderMirroringService.instance().reset();
};

export const mirrorAndWatchFolder = async (folderPath: string, folderId: string) => {
	let uuidCreationCount = 0;
	const createUuid = () => {
		const result = md5(uuidCreationCount++);
		return result;
	};
	return await FolderMirroringService.instance().mirrorFolder(folderPath, folderId, createUuid);
};
