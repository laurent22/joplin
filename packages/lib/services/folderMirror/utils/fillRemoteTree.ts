import shim from '../../../shim';
import ItemTree, { AddActionListener } from '../ItemTree';
import debugLogger from './debugLogger';
import statToItem from './statToItem';


const fillRemoteTree = async (baseFolderPath: string, remoteTree: ItemTree, addItemHandler: AddActionListener) => {
	const stats = await shim.fsDriver().readDirStats(baseFolderPath, { recursive: true });

	// Sort so that parent folders are visited before child folders.
	stats.sort((a, b) => a.path.length - b.path.length);

	for (const stat of stats) {
		const item = await statToItem(baseFolderPath, stat, remoteTree);
		if (!item) continue;

		if (remoteTree.hasId(item.id)) {
			debugLogger.warn('Removing ID from item with duplicated ID. Item: ', item.title, item.id, 'at', stat.path);
			delete item.id;
		}

		await remoteTree.addItemAt(stat.path, item, addItemHandler);
	}
};

export default fillRemoteTree;
