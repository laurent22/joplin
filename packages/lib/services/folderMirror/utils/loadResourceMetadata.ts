import { resourceMetadataExtension } from '../constants';
import shim from '../../../shim';
import * as yaml from 'js-yaml';
import { ResourceEntity } from '../../database/types';
import { basename, extname } from 'path';
import debugLogger from './debugLogger';
import { fromFilename as mimeFromFilename } from '../../../mime-utils';


const loadResourceMetadata = async (resourcePath: string): Promise<ResourceEntity> => {
	const metadataPath = `${resourcePath}${resourceMetadataExtension}`;

	const filename = basename(resourcePath);
	const extension = extname(resourcePath);
	const result: ResourceEntity = {
		mime: mimeFromFilename(filename),
		filename,
		title: filename.substring(0, filename.length - extension.length),
		// Internally, file_extension should not have a leading ".". Otherwise, the dot will
		// be repeated when saving.
		file_extension: extension.replace(/^\./, ''),
	};

	if (!await shim.fsDriver().exists(metadataPath)) {
		debugLogger.debug('Couldn\'t find metadata at path', metadataPath);
		return result;
	}

	const metadataString = await shim.fsDriver().readFile(metadataPath, 'utf8');
	const metadata = yaml.load(metadataString);
	if (typeof metadata !== 'object') {
		throw new Error('Wrong resource metadata format. Expected YAML object.');
	}

	if (!('id' in metadata) || typeof metadata.id !== 'string') {
		throw new Error('Expected metadata.id to be a string');
	}

	result.id = metadata.id;

	if ('title' in metadata) {
		result.title = `${metadata.title}`;
	}

	if ('ocr_text' in metadata) {
		result.ocr_text = `${metadata.ocr_text}`;
	}

	debugLogger.debug('Loaded metadata', result);
	return result;
};

export default loadResourceMetadata;
