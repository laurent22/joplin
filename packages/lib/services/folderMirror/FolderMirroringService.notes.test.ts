import { join } from 'path';
import Note from '../../models/Note';
import { createTempDir } from '../../testing/test-utils';
import createFilesFromPathRecord from '../../utils/pathRecord/createFilesFromPathRecord';
import { mirrorAndWatchFolder, resetFolderMirror, setUpFolderMirror, waitForNoteChange, waitForTestNoteToBeWritten } from './utils/test-utils';
import * as fs from 'fs-extra';
import Folder from '../../models/Folder';
import directoryToPathRecord from '../../utils/pathRecord/directoryToPathRecord';

describe('FolderMirroringService.notes', () => {
	beforeEach(async () => {
		await setUpFolderMirror();
	});

	afterEach(async () => {
		await resetFolderMirror();
	});

	//
	// IDs
	//

	test('should assign IDs to notes that lack them on an initial sync', async () => {
		const tempDir = await createTempDir();
		await createFilesFromPathRecord(tempDir, {
			'a.md': '---\ntitle: A test\n---',
			'b.md': '---\ntitle: Another test\n---\n\n# Content',
			'c.md': '---\ntitle: Another test\nid: 0123456789abcdef0123456789abcdef---\n\n# This note already has an ID',
			'test/foo/c.md': 'Another note',
		});
		await mirrorAndWatchFolder(tempDir, '');
		expect(await directoryToPathRecord(tempDir)).toMatchSnapshot('should assign IDs to notes that lack them on an initial sync');

		await fs.writeFile(join(tempDir, 'new.md'), 'New file with no ID');
		await waitForTestNoteToBeWritten(tempDir);

		expect(await directoryToPathRecord(tempDir)).toMatchSnapshot('should assign IDs to new notes written while watching');
	});

	//
	// Syncing changes
	//

	test('should modify items locally when changed in a watched, non-empty remote folder', async () => {
		const tempDir = await createTempDir();
		await createFilesFromPathRecord(tempDir, {
			'a.md': '---\ntitle: A test\n---',
			'b.md': '---\ntitle: Another test\n---\n\n# Content',
			'test/foo/c.md': 'Another note',
		});
		await mirrorAndWatchFolder(tempDir, '');

		expect(await Note.loadByTitle('A test')).toMatchObject({ body: '', parent_id: '' });

		const changeListener = waitForNoteChange(note => note.body === 'New content');
		await fs.writeFile(join(tempDir, 'a.md'), '---\ntitle: A test\n---\n\nNew content', 'utf8');
		await changeListener;
		await waitForTestNoteToBeWritten(tempDir);

		expect(await Note.loadByTitle('A test')).toMatchObject({
			body: 'New content', parent_id: '',
		});
	});

	test('should move notes when moved in a watched folder', async () => {
		const tempDir = await createTempDir();
		await createFilesFromPathRecord(tempDir, {
			'a.md': '---\ntitle: A test\n---',
			'test/foo/c.md': 'Another note',
		});
		await mirrorAndWatchFolder(tempDir, '');


		const testFolderId = (await Folder.loadByTitle('test')).id;
		const noteId = (await Note.loadByTitle('A test')).id;

		await fs.move(join(tempDir, 'a.md'), join(tempDir, 'test', 'a.md'));

		await waitForTestNoteToBeWritten(tempDir);

		const movedNote = await Note.loadByTitle('A test');
		expect(movedNote).toMatchObject({ parent_id: testFolderId, id: noteId });
	});
});
