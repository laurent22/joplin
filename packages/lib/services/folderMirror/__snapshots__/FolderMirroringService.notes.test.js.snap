// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FolderMirroringService.notes should assign IDs to notes that lack them on an initial sync: should assign IDs to new notes written while watching 1`] = `
{
  "a.md": "---
title: A test
id: 93b885adfe0da089cdf634904fd59f71
---

",
  "b.md": "---
title: Another test
id: 55a54008ad1ba589aa210d2629c1df41
---

# Content",
  "c.md": "---
title: Another test
id: 9e688c58a5487b8eaf69c9e1005ad0bf
---

",
  "new.md": "New file with no ID",
  "test/.folder.yml": "title: test
id: 8666683506aacd900bbd5a74ac4edf68
",
  "test/foo/.folder.yml": "title: foo
id: ec7f7e7bb43742ce868145f71d37b53c
",
  "test/foo/c.md": "---
title: c
id: 8bb6c17838643f9691cc6a4de6c51709
---

Another note",
}
`;

exports[`FolderMirroringService.notes should assign IDs to notes that lack them on an initial sync: should assign IDs to notes that lack them on an initial sync 1`] = `
{
  "a.md": "---
title: A test
id: 93b885adfe0da089cdf634904fd59f71
---

",
  "b.md": "---
title: Another test
id: 55a54008ad1ba589aa210d2629c1df41
---

# Content",
  "c.md": "---
title: Another test
id: 9e688c58a5487b8eaf69c9e1005ad0bf
---

",
  "test/.folder.yml": "title: test
id: 8666683506aacd900bbd5a74ac4edf68
",
  "test/foo/.folder.yml": "title: foo
id: ec7f7e7bb43742ce868145f71d37b53c
",
  "test/foo/c.md": "---
title: c
id: 8bb6c17838643f9691cc6a4de6c51709
---

Another note",
}
`;
